// @nrwl/workspace/src/core/project-graph/build-dependencies/explicit-package-json-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildExplicitPackageJsonDependencies = void 0;
const file_utils_1 = require("../../file-utils");
const devkit_1 = require("@nrwl/devkit");
function buildExplicitPackageJsonDependencies(workspace, graph, filesToProcess) {
    const res = [];
    Object.keys(filesToProcess).forEach((source) => {
        Object.values(filesToProcess[source]).forEach((f) => {
            if (isPackageJsonAtProjectRoot(graph.nodes, f.file)) {
                processPackageJson(source, f.file, graph, res);
            }
        });
    });
    return res;
}
exports.buildExplicitPackageJsonDependencies = buildExplicitPackageJsonDependencies;
function isPackageJsonAtProjectRoot(nodes, fileName) {
    return Object.values(nodes).find((projectNode) => (projectNode.type === 'lib' || projectNode.type === 'app') &&
        devkit_1.joinPathFragments(projectNode.data.root, 'package.json') === fileName);
}
function processPackageJson(sourceProject, fileName, graph, collectedDeps) {
    try {
        const deps = readDeps(devkit_1.parseJson(file_utils_1.defaultFileRead(fileName)));
        deps.forEach((d) => {
            // package.json refers to another project in the monorepo
            if (graph.nodes[d]) {
                collectedDeps.push({
                    sourceProjectName: sourceProject,
                    targetProjectName: d,
                    sourceProjectFile: fileName,
                });
            }
        });
    }
    catch (e) {
        if (process.env.NX_VERBOSE_LOGGING === 'true') {
            console.log(e);
        }
    }
}
function readDeps(packageJsonDeps) {
    var _a, _b;
    return [
        ...Object.keys((_a = packageJsonDeps === null || packageJsonDeps === void 0 ? void 0 : packageJsonDeps.dependencies) !== null && _a !== void 0 ? _a : {}),
        ...Object.keys((_b = packageJsonDeps === null || packageJsonDeps === void 0 ? void 0 : packageJsonDeps.devDependencies) !== null && _b !== void 0 ? _b : {}),
    ];
}
